File: advent_of_code_2023/day15/Main.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Control.Applicative ((<|>))
import Control.Monad.ST (runST)
import Data.Attoparsec.Text
import Data.Char (ord)
import Data.Foldable (for_, traverse_)
import Data.Text (Text)
import Data.Text qualified as T
import Data.Text.IO qualified as T
import Data.Vector (Vector)
import Data.Vector qualified as V
import Data.Vector.Mutable qualified as M

hash :: Text -> Int
hash = T.foldl' (\h c -> (17 * (h + ord c)) `rem` 256) 0

part1 :: Text -> Int
part1 = sum . fmap hash . T.split (== ',')

data Op = Dash | Equal Int deriving (Show)
type Step = (Text, Op)

readSteps :: Text -> [Step]
readSteps = either (error "Bad parse") id . parseOnly ((step `sepBy1'` ",") <* endOfInput)
  where
    step = (,) <$> label <*> op
    label = T.pack <$> many1' letter
    op = (Dash <$ "-") <|> (Equal <$> ("=" *> decimal))

set :: (Text, Int) -> [(Text, Int)] -> [(Text, Int)]
set (a, m) [] = [(a, m)]
set (a, m) ((b, n) : bs)
    | a == b = (a, m) : bs
    | otherwise = (b, n) : set (a, m) bs

perform :: [Step] -> Vector [(Text, Int)]
perform ss = runST $ do
    b <- V.unsafeThaw $ V.replicate 256 []
    for_ ss $ \(label, op) ->
        let f = case op of
                Dash -> filter ((/= label) . fst)
                Equal i -> set (label, i)
         in M.unsafeModify b f (hash label)
    V.unsafeFreeze b

part2 :: Text -> Int
part2 =
    V.sum
        . V.imap (\i b -> sum $ zipWith (\j (_l, k) -> (1 + i) * j * k) [1 ..] b)
        . perform
        . readSteps

main :: IO ()
main = do
    input <- T.readFile "input.txt"
    traverse_ (print . ($ input)) [part1, part2]

File: advent_of_code_2023/day15/src/main.rs
use indexmap::IndexMap;
use std::fs;

fn hash(s: &str) -> usize {
    s.chars().fold(0, |h, c| ((h + c as usize) * 17) % 256)
}

fn part1(input: &str) -> usize {
    input.split(',').map(hash).sum()
}

fn part2(s: &str) -> usize {
    let mut boxes = vec![IndexMap::<&str, usize>::new(); 256];
    for step in s.split(',') {
        let h = hash(step.trim_end_matches(|c: char| c == '-' || c == '=' || c.is_numeric()));
        if step.ends_with('-') {
            let _ = boxes[h].shift_remove(&step[0..step.len() - 1]);
        } else {
            let (l, n) = step.split_once("=").unwrap();
            boxes[h].insert(l, n.parse().unwrap());
        }
    }
    boxes
        .iter()
        .enumerate()
        .map(|(i, b)| {
            b.values()
                .enumerate()
                .map(|(j, l)| (i + 1) * (j + 1) * l)
                .sum::<usize>()
        })
        .sum()
}

fn main() {
    let input = fs::read_to_string("input.txt").expect("Should be able to read input");
    println!("{}", part1(&input));
    println!("{}", part2(&input));
}
